genomepy/files.py

```diff
@@ -122,21 +146,126 @@ def gunzip_and_name(fname: str) -> Tuple[str, bool]:
     fname: str
         path to file
 
+    outfile : str, optional
+        Name of the output file.
+
+    concat : bool, optional
+        If zip or tar.gz contains multiple files, concatenate them.
+
     Returns
     -------
     tuple
         fname, str
             up to date filename
-        gzip, bool
-            whether the file was gunzipped
     """
+    if fname.endswith(".gz"):
+        return extract_gzip(fname, outfile=outfile)
+    elif fname.endswith(
+        ".zip",
+    ):
+        return extract_zip(fname, outfile=outfile, concat=concat)
+    elif fname.endswith(".tgz") or fname.endswith(".tar.gz"):
+        return extract_tarball(fname, outfile=outfile)
+
+
+def extract_tarball(fname, outfile=None, concat=True) -> Union[str, None]:
+    """Convert tar of multiple FASTAs to one file."""
+    fnames = []
+    # Extract files to temporary directory
+    tmp_dir = mkdtemp(dir=os.path.dirname(outfile))
+    with tarfile.open(fname) as tar:
+        tar.extractall(path=tmp_dir)
+    for root, _, files in os.walk(tmp_dir):
+        fnames += [os.path.join(root, fname) for fname in files]
+
+    if len(fnames) > 1 and not concat:
+        raise ValueError("tarball contains multiple files, but concat not specified!")
+
+    # Concatenate (also works woth one file)
+    with open(outfile, "w") as out:
+        for infile in fnames:
+            for line in open(infile):
+                out.write(line)
+
+    rm_rf(tmp_dir)
+
+    return outfile
+
+
+def extract_gzip(fname: str, outfile: Optional[str] = None) -> Union[str, None]:
+    """
+    Gunzips the file if gzipped.
+
+    Also works on bgzipped files.
+
+    Parameters
+    ----------
+    fname: str
+        path to file
+
+    outfile : str, optional
+        Name of the output file.
+
+    Returns
+    -------
+    tuple
+        fname, str
+            up to date filename, or None if failed
+    """
+    if not outfile:
+        outfile = fname[:-3]
+
     if fname.endswith(".gz"):
         with gzip.open(fname, "rb") as f_in:
-            with open(fname[:-3], "wb") as f_out:
+            with open(outfile, "wb") as f_out:
                 shutil.copyfileobj(f_in, f_out)
         os.unlink(fname)
-        return fname[:-3], True
-    return fname, False
+        return outfile
+
+
+def extract_zip(
+    fname: str, outfile: Optional[str] = None, concat: bool = False
+) -> Union[str, None]:
+    """
+    Unzips the file if zipped.
+
+    Parameters
+    ----------
+    fname: str
+        path to file
+
+    outfile : str, optional
+        Name of the output file.
+
+    concat : bool, optional
+        If zip or tar.gz contains multiple files, concatenate them.
+    Returns
+    -------
+    tuple
+        fname, str
+            up to date filename
+    """
+    if not outfile:
+        outfile = fname[:-4]
+
+    with ZipFile(fname, "r") as fzip:
+        with TemporaryDirectory() as tmpdir:
```

Comment:
`with ZipFile(fname, "r") as fzip, TemporaryDirectory() as tmpdir:`

just cleans up the indenting a bit